feat(sdk): Add C# SDK with module support

pjmagee opened the pull request:
experimental C# SDK with module support

- Attributes for Function with cache, Check, DefaultValue, Ignore, Field, Object, Name for alias support
- Defaults & optional values supported by C# syntax (string? value) or (string value = "hello")
- Constructor support with custom factory [Constructor] like python SDK `create()` for async
- Makes use of Roslyn analyzers for ensuring users of SDK use the API correctly, with IDE/Compiler hints,warnings, e.g. C# class not matching Dagger.json module name, or incorrectly using Attributes
- Makes use of .editorconfig to enable or override compiler/IDE C# analyzers
- .NET 10 long term support - officially released on November 11, 2025, and is supported for three years until November 14, 2028

- preview nuget package published at https://www.nuget.org/packages/DaggerIO/ will require Dagger team to create nuget.org account with API Key to publish update/packages but I followed existing sdks on version/bumping - happy for changes/reviews of any gaps (nuget package is NOT required for dagger modules, as the sdk is vendored with the generated code)

TomChv commented on sdk/csharp/src/Dagger.SDK.CodeFixes/AddDefaultPathAttributeCodeFixProvider.cs:
```diff
@@ -0,0 +1,91 @@
+        );
+    }
+
+    private async Task<Document> AddDefaultPathAttributeAsync(
```
I'm curious about that one: I'm not sure I understand the purpose of this CodeFix, why adding a default path on directory argument?

pjmagee replied:
e.g +defaultPath only works on `dagger.Directory` so this is to show the user they can add a default Path. It could also only be a hint.

TomChv replied:
Oh okay, that's a great suggestion then!

TomChv commented on sdk/csharp/src/Dagger.SDK/Attributes/FieldAttribute.cs:
```diff
@@ -0,0 +1,27 @@
+    AllowMultiple = false,
+    Inherited = false
+)]
+public class FieldAttribute : Attribute
```
`Field` is not really a thing on dagger because it doesn't really exist in `GraphQL`, everything is a function.

As decided and applied in https://github.com/dagger/dagger/pull/7701, `field` should be  decorated as `func` since these are GQL accesor.

So is it possible to use the `Function` attribute for fields?

pjmagee replied:
Ah, I was following ts sdk. Its possible to use the existing `Function` attribute. I will this as a todo to fix for this PR

pjmagee replied:
this part is now done

TomChv reviewed (commented):
:warning: I'm far from being a C# expert so some of my questions may look out of context due to my lack of knowledge.

That review is a first look, but there will be more coming next week.

The code looks very clean overall, thank you for making it easy to review!
Please rebase whenever you can, the repo is moving fast :)

Here's a summary of some tests I did:
- [x] Initialize a csharp project
- [x] Initialize a csharp project in a sub directory
- [x] Create and call check
- [x] Create and call function

About the performances, it's decently fast and usable, I'll not worry to much about optimization for now, we can follow up later. 

The goal is to merge that one during January, so it gives us a bit of time to prepare the ground and sets the current dotnet SDK as deprecated.

TomChv commented on sdk/csharp/src/Dagger.SDK.CodeFixes/AddFunctionAttributeCodeFixProvider.cs:
```diff
@@ -0,0 +1,76 @@
+        );
+    }
+
+    private async Task<Document> AddFunctionAttributeAsync(
```
I think it's okay as a lightbulb, so it's easy to expose a function :)

TomChv commented:
> Regarding the Analyzers/Codefixes - They are also entirely optional and can be completely disabled, my idea behind adding them was to add some 'rails' or quick fixes. Though they could be considered maybe opinionated, or not 'complete'.

I think the analyzers/codefixes are a great idea! I wish I could add that to other SDKs too!

pjmagee commented on sdk/csharp/src/Dagger.SDK.CodeGen/Code/CodeRenderer.cs:
```diff
@@ -0,0 +1,739 @@
+
+            if (isAsync)
+            {
+                methodName = $"{methodName}Async";
```
Yes, I also did something a bit funky with interfaces. I am tempted to remove the async suffix AND remove the implicit interace trimming of `I` that usually is seen across 99.9% of C# Code. Maybe I can add something like `[Interface(Name = "CustomName")]` and then that custom name is used to register the interface. Might be a bit cleaner.... and also removing hiding implicit logic of stripping any 'I' which now looking at it, i don't like.

eunomie commented:
> @dagger/maintainers so should we merge this?

IIRC one of the question was (maybe on discord?) how this could replace the already existing dotnet sdk. Or maybe we want to keep both?
I don't know if @TomChv you discussed a more detailed plan on that aspect?

pjmagee commented on core/sdk/loader.go:
```diff
@@ -138,6 +138,8 @@ func (l *Loader) namedSDK(
 	case sdkElixir:
 		return l.SDKForModule(ctx, root, &core.SDKConfig{Source: "github.com/dagger/dagger/sdk/elixir" + sdkSuffix, Config: sdk.Config, Experimental: sdk.Experimental}, nil)
+	case sdkCSharp:
+		return l.SDKForModule(ctx, root, &core.SDKConfig{Source: "github.com/pjmagee/dagger/sdk/csharp@csharp/experimental" + sdkSuffix, Config: sdk.Config, Experimental: sdk.Experimental}, nil)
```
This would need updating actually once merged

TomChv commented:
If you can make the PR green, I would be happy to give it another review so we can merge that

pjmagee committed:
feat(sdk): Add C# SDK with module support

This introduces a complete C# SDK implementation for Dagger with full module development support.

Core Features:
- Full Dagger API client with fluent chainable interface
- Module runtime with attribute-based type registration
- Support for objects, interfaces, enums, and functions
- Constructor-based dependency injection with async factory methods
- Comprehensive attribute system for API customization

Attributes:
- [Object] - Mark classes as Dagger modules
- [Interface] - Define interface types
- [Function] - Expose methods and properties
- [Constructor] - Alternative static factory constructors
- [Name] - Custom parameter naming (e.g., for keyword conflicts)
- [DefaultPath] - Load Directory/File from module context
- [Ignore] - Exclude patterns from Directory arguments
- [Deprecated] - Mark parameters/functions as deprecated
- [Check] - Define validation functions
- [Enum]/[EnumValue] - Custom enum types

Module System:
- Automatic type discovery and registration
- GraphQL schema generation from C# types
- State management with JSON serialization
- Support for Dagger types (Container, Directory, File, etc.)
- Interface implementation with dynamic proxies

Developer Experience:
- Roslyn analyzers for compile-time validation
- Code fixes for common patterns
- Comprehensive examples demonstrating all features
- Integration with existing Dagger CLI tooling

Signed-off-by: pjmagee <pjmagee@users.noreply.github.com>
